sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -296,6 +296,46 @@ case class UpdateTable(
   override def children: Seq[LogicalPlan] = table :: Nil
 }
 
+/**
+ * The logical plan of the MERGE INTO command that works for v2 tables.
+ */
+case class MergeIntoTable(
+    targetTable: LogicalPlan,
+    sourceTable: LogicalPlan,
+    mergeCondition: Expression,
+    matchedActions: Seq[MergeAction],
+    notMatchedActions: Seq[MergeAction]) extends Command with SupportsSubquery {
+  override def children: Seq[LogicalPlan] =
+    Seq(targetTable, sourceTable) ++ Nil
+  override def output: Seq[Attribute] = Seq.empty
+  override lazy val resolved: Boolean = expressions.forall(_.resolved) && childrenResolved &&
+      matchedActions.forall(_.resolved) && notMatchedActions.forall(_.resolved)
+}
+
+sealed abstract class MergeAction(
+    condition: Option[Expression]) extends Expression with Unevaluable {
+  override def foldable: Boolean = false
+  override def nullable: Boolean = false
+  override def dataType: DataType = null
+  override def children: Seq[Expression] = condition.toSeq
+}
+
+case class DeleteAction(deleteCondition: Option[Expression]) extends MergeAction(deleteCondition)
+
+case class UpdateAction(
+    updateCondition: Option[Expression],
+    columns: Seq[Expression],
```

Comment:
It's a bit tricky as we need to guarantee `columns` and `values` have the same length. How about
```
case class Assignment(column: Expression, value: Expression)
case class UpdateAction(condition: Option[Expression], assignments: Seq[Assignment])
```

Reply:
Then we can reuse the analysis code between update and insert action to resolve `Assignment`

Reply:
If `assignments` is Nil, then it means `UPDATE SET *` or `INSERT *`

Reply:
Done of replacing `columns` and `values` with `Seq[Assignment]`. Please have a look.